### ISSUE TITLE
API design for html text translation

### ISSUE THREAD

**Issue by qianqianzhu** [browsermt/bergamot-translator#238]
Current Service classes (including `BlockingService` and `AsyncService`) only accept source text. (BTW, why is there inconsistency between `BlockingService` and `AsyncService` in terms of receiving source text? In `BlockingService`, source text is passed as `std::vector<std::string>` while as `std::string` in `AsyncService`.) 

To support html text translation, an additional input, i.e., tag positions as a `std::vector<ByteRange>` needed to be received. It becomes a critical part for integration and continuing work of https://github.com/browsermt/bergamot-translator/pull/199. As there are bindings and dependencies relying on Service classes, I opened this issue for feedback and discussion (@kpu @jerinphilip @XapaJIaMnu @abhi-agg @andrenatal)

The basic idea as we promised to Mozilla is:

from the browser side, we would expect the entire text (no matter whether it contains a phrase, single sentence or multiple sentences) where all the tags are removed and one `std::vector<ByteRange>`where contains tag positions at char-level. Then from us, you would expect the translated text as whole (without tags) and one `std::vector<ByteRange>` where contains tag positions at char-level in the translated text (in the same order as in the original text so you render them back).

Since this tag alignment feature will be mainly used by the browser, my proposal would be extending current `BlockingService` to accept additional parameters (of course `AsyncService` can also be extended if necessary):
```
std::vector<Response> translateMultiple(std::shared_ptr<TranslationModel> translationModel, 
                                       std::vector<std::string> &&source, 
                                       const ResponseOptions &responseOptions, 
                                       const std::vector<ByteRange> &tagPositonSource={});
```
And in `Response` class, a new member will be added `std::vector<ByteRange> tagPositonTarget` for tag positions (at char-level) in the translated text

Feel free to comment.

**Comment by jerinphilip** [comment]
> The integration/extension for AsyncService class can come later once the tag processor is mature.

As a general guideline, target `AsyncService` first. `BlockingService` is single-threaded and blocking, like the name says. My reasoning is:

1. 1-thread takes a long time to run on our typical WNGT20 tests which benchmark for speed. For example, if we want to measure how much overhead your tag-tree adds over a span of 1M requests, we have much faster turnaround with `AsyncService`. If we want to run across all random-walk generated HTML trees, we can find corner cases/search space much faster with `AsyncService` using 48 threads on `var` compared to 1 thread with `BlockingService`.

2. It's possible to convert `AsyncService` to `BlockingService`, but not the other way around because thread-safety issues will arise. 


> I am not sure why AsyncService accepts only std::string though

Because it's asynchronous, all you have to do is create a vector, allow the callback to write on to the vector, and setup a barrier to fire to get things to parity with `BlockingService` (This is how `BlockingService::translateMultiple` is implemented without a barrier because duh, single-threaded). Perhaps a convenience function can be offered for this case, but a decent client would use the translation immediately as available, not compile it into vectors, pack it and unpack it again. There's no consumer for such a function embedded on to `AsyncService` given we already provision `BlockingService` for Mozilla.

**Comment by qianqianzhu** [comment]
The API integration of tag alignment is highlighted in [PR#241](https://github.com/browsermt/bergamot-translator/pull/241):
Both `AsyncService` and `BlockingService` Class are modified.

For `BlockingService`, an additional parameter is added to pass tag positions (at char-level) in source text as `std::vector<TagPositions> &&tagPositionSources`:
https://github.com/browsermt/bergamot-translator/blob/e5c0837d748bdc8e9ae62b2ff8bba159e8074bb5/src/translator/service.h#L53-L55
The tag positions (at char-level) in the translated text are returned as a new member `response.target.tagPositions`in Response Class:
https://github.com/browsermt/bergamot-translator/blob/e5c0837d748bdc8e9ae62b2ff8bba159e8074bb5/src/tests/apps.cpp#L195-L202

**Comment by kpu** [comment]
There's a design flaw in the tag API regarding zero-length tags.  Consider "&lt;span&gt;&lt;/span&gt;&lt;b&gt;&lt;/b&gt;".  It has the same encoding "&lt;span&gt;&lt;b&gt;&lt;/b&gt;&lt;/span&gt;" namely `{ByteRange{0,0}, ByteRange{0,0}}`

**Comment by qianqianzhu** [comment]
> In HTML, using a closing tag on an empty element is usually invalid. For example, <input type="text"></input> is invalid HTML.

Ref https://developer.mozilla.org/en-US/docs/Glossary/Empty_element

**Comment by kpu** [comment]
My reading of that document is that &lt;br&gt; and similar tags like &lt;hr&gt; are not allowed to have content inside them.  

It does not say that `span` must have text in it.

**Comment by qianqianzhu** [comment]
What are you suggesting then?

**Comment by kpu** [comment]
Either we take in tags in some simplified form or maybe a sequence of tag opening and closing events where the top bit is 0 for opening and 1 for closing?  

So "&lt;span&gt;&lt;/span&gt;&lt;b&gt;&lt;/b&gt;" becomes 0, 1<<31, 0, 1 << 31

**Comment by kpu** [comment]
But then that doesn't have a nice way of communicating back to the client when tags are permuted.

**Comment by jerinphilip** [comment]
If we don't go for the flat structure, tree structure exchanges can handle nesting versus sequential, yes? I would prefer such a setting in JS land compared to bit-shift representations. We can attach identifiers along with the ByteRange as well to recompile when at target.

**Comment by qianqianzhu** [comment]
You are absolutely right. Can you suggest a better way to pass tag positions?

**Comment by kpu** [comment]
I am wondering if I should just get over my opposition to parsing HTML in C++ and use this BSD-licensed code from https://www.codeproject.com/Articles/14076/Fast-and-Compact-HTML-XML-Scanner-Tokenizer?msg=1536730#xx1536730xx 
which warc2text is already using https://github.com/bitextor/warc2text/blob/master/src/xh_scanner.hh

**Comment by qianqianzhu** [comment]
Closed as this issue was resolved in [PR#253](https://github.com/browsermt/bergamot-translator/pull/253).
